src/api/forms/insert.ts

```diff
@@ -1,18 +1,36 @@
-import { InsertForm, InsertFormType, filterRecordType } from './base';
+import { InsertForm, InsertFormType, isRecordType, filterRecordType, InsertFormRecord } from './base';
 import * as db from '../db';
 
-export async function insertForm(form: InsertForm) {
+export async function insertForm(form: InsertForm) : Promise<InsertFormRecord | void> {
 	// map local (within `form`) ids to created database ids
 	const contentIdMap: Map<number, number> = new Map();
 	const relationIdMap: Map<number, number> = new Map();
 	const citationIdMap: Map<number, number> = new Map();
+	const attributionIdMap: Map<number, number> = new Map();
 
 	for (const record of filterRecordType(form, InsertFormType.CONTENT)) {
 		const { id, ...value } = record.value;
 
-		// Insert content node & update map with returned id
-		const newId = await db.insertContent(value);
-		contentIdMap.set(id, newId);
+		if (id < 0) {
+			// Insert content node & update map with returned id
+			const newId = await db.insertContent(value);
+			contentIdMap.set(id, newId);
+		} else {
+			// Insert revision for updating content
+			await db.insertRevision({
+				type: "content",
+				link_modifies: id,
+				created: new Date(),
+				content: {
+					// TODO: add a property for the user that created the revision
+					from: (await db.getContent(id)) as never,
+					to: value as never,
+				},
```

Comment:
[`never`](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#any-unknown-object-void-undefined-null-and-never-assignability) is a part of TypeScript - so it doesn't end up in the compiled JS code, but it does inform TypeScript's type checker about how it should treat those values. (the `as` keyword is equivalent to a type cast)

In this case, the values in the "content" column are all untyped, and TypeScript doesn't like there being an index type on the `Content` object that it can't parse through, so it is necessary to ignore those typings to prevent it from marking that as an error.

Reply:
Ah, okay, so giving a variable the type `never` is basically just a way of saying "this is actually untyped, don't panic"?

Reply:
Yep, exactly! (it's more of a "this will never occur", but that has the same effect)